### PR TITLE
Organization and Region rollup in Dashboard and Inventory

### LINKED ISSUE #509: UI: Create new MUI Side Drawer (CRASM-545)
# 🐛 Summary #
- Directives to improve the Organizational drop down menu and create a new Regional drop down menu necessitate a refactoring of the existing filter drawer in Inventory.
- This filtering by Region and/or Organization will affect data displayed both in Overview and Inventory.
- To that end we will need a new Side Drawer as follows:
- Create a new MUI based Side Drawer that combines existing Search, existing filters, and Saved Search with the new Regional and Organizational drop down menus.
- This Side Drawer will be accessible via tab or button from both the Overview and Inventory page.
- It will be persistent in desktop.
- It will be triggered by toggle in mobile.

## Any helpful log output or screenshots ##

image
image
image
image

Add any screenshots of the problem here.

### PR DESCRIPTION
**Pull request by jayjaybunce**
#  #

## 🗣 Description ##

- Added search functionality across organizations with filtering for regions via /organizations/search
- Updated domain search to accept filters for regions
- Added region list functionality for frontend filtering via /regions
- Updated Layout component to initially load default filter values based on user type
- Updated Header to display drawer toggle
- Implemented FilterDrawerV2 which includes a mobile and desktop version
- Updated stats endpoint handler to support region filtering
- Added utils to enable organizations search index syncing
- Updated useUserActivity hook to reduce unnecessary rendering

## 💭 Motivation and context ##

Resolves [CRASM-403](https://maestro.dhs.gov/jira/browse/CRASM-403)

## 🧪 Testing ##

After pulling down branch:
1. Stop any running containers
2. `projectroot` => run `npm run start`
3. `backeng/src/tasks/syncdb.ts` => change `NUM_SAMPLE_ORGS` value to 100
4. `backend`  => run `npm run syncdb -- -d dangerouslyforce`
5. `backend` => run `npm run syncdb -- -d populate`
6. Login to app locally and test.

Testing parameters used:
1. Filter drawer is accessible on `/` and `/inventory`
2. All roles => Filter drawer is populated with default filters for region and organization
3. Standard user => Cannot alter these filters 
4. Region admin user => Able to alter region filters and implement new organization filters via the search input
5. Global admin + Global view => Able to alter all filters

Testing Inventory
1. Altering region and organization filter will either narrow or broaden the results, test at your own discretion

Testing Overview
1. Altering any filters will either narrow or broaden the results, test at your own discretion

*Note - Regional Admin Toggle*
- Within `frontend/src/hooks/useUserTypeFilters.ts` there is a variable `REGIONAL_USER_CAN_SEARCH_OTHER_REGIONS`
- If `true` => Regional Admin will be able to interact with region filters and in doing so see results for stats and domains in other regions
- If `false` => Region Admin will not be able to change region filters and are restricted to stats and domains exclusively in their own region

### Overview
Screenshot 2024-08-13 at 11 35 05 AM

### Drawer => Overview => Standard user
Screenshot 2024-08-13 at 11 36 23 AM

### Drawer => Inventory => Standard user
Screenshot 2024-08-13 at 11 36 56 AM

### Drawer => Overview => RegionalAdmin + GlobalAdmin + GlobalView
Screenshot 2024-08-13 at 11 38 42 AM

### Drawer => Inventory => RegionalAdmin + GlobalAdmin + GlobalView
Screenshot 2024-08-13 at 11 38 14 AM

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [ ] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [ ] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Create a release.

### CODE REVIEW

**Review by ameliav — COMMENTED** [review]
To fix the frontend Github checks, you'll need to run pre-commit and npm test:

```
cd /frontend
pre-commit run --all-files
npm test
u
```
and push up the file changes

File: backend/src/api/regions.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { connectToDatabase, Organization } from '../models';
```
Comment (chrtorres) [inline backend/src/api/regions.ts]: There's an unused import for Organization

**Review by ameliav — COMMENTED** [review]
The frontend github checks look good now.

There's a couple new warnings in the console. They don't break functionality so they don't need to be fixed right away but should be tracked- you can write in "TODOs" in the code to fix them in the future.

EDIT- I ended up fixing the warnings in the console. There is still an Okta error- however, it doesn't seem to break functionality and I am not sure if that is related to the new code or was there before.

File: backend/src/api/stats.ts [hunk 2]
```diff
@@ -1,4 +1,11 @@
-import { ValidateNested, IsOptional, IsObject, IsUUID } from 'class-validator';
+import {
+  ValidateNested,
+  IsOptional,
+  IsObject,
+  IsUUID,
+  isUUID,
```
Comment (chrtorres) [inline backend/src/api/stats.ts]: Unused or erroneous "isUUID"

File: backend/src/api/stats.ts [hunk 3]
```diff
@@ -31,13 +38,20 @@ interface Stats {
 }
```
Comment (chrtorres) [inline backend/src/api/stats.ts:33]: isGlobalViewAdmin and getOrgMemberships are no longer being used, these should be removed.

File: backend/src/tasks/es-client.ts [hunk 4]
```diff
@@ -1,13 +1,22 @@
 import { Client } from '@elastic/elasticsearch';
-import { Domain, Webpage } from '../models';
+import { Domain, Organization, Webpage } from '../models';
```
Comment (chrtorres) [inline backend/src/tasks/es-client.ts]: Unused import "Webpage"

File: backend/src/tasks/search-sync-domains.ts [hunk 5]
```diff
@@ -1,4 +1,10 @@
-import { Domain, connectToDatabase, Vulnerability, Webpage } from '../models';
+import {
+  Domain,
+  connectToDatabase,
+  Vulnerability,
+  Webpage,
+  Organization
+} from '../models';
```
Comment (chrtorres) [inline backend/src/tasks/search-sync-domains.ts]: Unused Imports: Vulnerability, Webpage, Organization

File: backend/src/tasks/search-sync-orgs.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import {
+  Domain,
+  connectToDatabase,
+  Vulnerability,
+  Webpage,
+  Organization
+} from '../models';
+import { CommandOptions } from './ecs-client';
```
Comment (chrtorres) [inline backend/src/tasks/search-sync-orgs.ts]: Unused imports: 
- Domain
- connectToDatabase
- Vulnerability
- Webpage
- CommandOptions

**Review by chrtorres — CHANGES_REQUESTED** [review]
There are some lingering unused imports in the backend directory. There is some weird resizing on the My account pages, but functionality is not impeded by this and can be addressed in a separate issue. Once the unused imports are addressed I can approve this.

File: backend/src/api/stats.ts (hunk 2 above)
Comment (ameliav) [reply backend/src/api/stats.ts]: e578861fa6d61912c7f0d44cb8e69de7011e51d0

File: backend/src/api/stats.ts (hunk 3 above)
Comment (ameliav) [reply backend/src/api/stats.ts:33]: Resolved e578861fa6d61912c7f0d44cb8e69de7011e51d0

File: backend/src/tasks/es-client.ts (hunk 4 above)
Comment (ameliav) [reply backend/src/tasks/es-client.ts]: Resolved e578861fa6d61912c7f0d44cb8e69de7011e51d0

File: backend/src/tasks/search-sync-domains.ts (hunk 5 above)
Comment (ameliav) [reply backend/src/tasks/search-sync-domains.ts]: Resolved e578861fa6d61912c7f0d44cb8e69de7011e51d0

File: backend/src/tasks/search-sync-orgs.ts (hunk 6 above)
Comment (ameliav) [reply backend/src/tasks/search-sync-orgs.ts]: Resolved e578861fa6d61912c7f0d44cb8e69de7011e51d0

File: backend/src/api/regions.ts (hunk 1 above)
Comment (ameliav) [reply backend/src/api/regions.ts]: Resolved [6ca0ca7](https://github.com/cisagov/XFD/pull/526/commits/6ca0ca7e6d0ade414828e350dc12b4b9e505503c)

**Review by ameliav — CHANGES_REQUESTED** [review]
From the frontend perspective github checks are passing now, console warnings are taken care of, and functionality works.

image
However, there is a breakpoint before mobile that squishes the card headers.

**Review by ameliav — APPROVED** [review]
LGTM from the front-end local build perspective!

**Review by chrtorres — APPROVED** [review]

**Review by schmelz21 — APPROVED** [review]

**Pull request merged**